port/services.js

```diff
@@ -0,0 +1,69 @@
+import { PortWithRequests } from './requests.js';
+
+const typeNameValidator = /^[a-zA-Z]+[a-zA-Z-]*$/;
+
+export class PortWithServices extends PortWithRequests {
+
+	constructor(endPoint, targetOrigin, options) {
+		super(endPoint, targetOrigin, options);
+	}
+
+	async getService(serviceType, version) {
+		if (!this._isConnected) {
+			throw new Error('Cannot getService() before connect() has completed');
+		}
+
+		const serviceVersionPrefix = `service:${serviceType}:${version}`;
+		const me = this;
+
+		function createProxy(methodNames) {
+			function createProxyMethod(name) {
+				return function() {
+					const args = new Array(arguments.length + 1);
+					args[0] = `${serviceVersionPrefix}:${name}`;
+					for (let i = 0; i < arguments.length; ++i) {
+						args[i + 1] = arguments[i];
+					}
+
+					return me.request.apply(me, args);
+				};
+			}
+
+			const proxy = {};
+			methodNames.forEach(function(name) {
+				proxy[name] = createProxyMethod(name);
+			});
+			return proxy;
+		}
+
+		const methodNames_1 = await this.request(serviceVersionPrefix);
+		return createProxy(methodNames_1);
+	}
+
+	registerService(serviceType, version, service) {
+		if (this._isConnected) {
+			throw new Error('Register services before connecting');
+		}
+
+		if (!typeNameValidator.test(serviceType)) {
+			throw new Error(`Invalid service type ${serviceType}`);
+		}
+
+		const serviceVersionPrefix = `service:${serviceType}:${version}`;
+
+		const methodNames = Object
+			.keys(service)
+			.filter(k => {
+				return typeof service[k] === 'function';
+			});
+
+		this.onRequest(serviceVersionPrefix, methodNames);
+
+		methodNames.forEach(name => {
+			this.onRequest(`${serviceVersionPrefix}:${name}`, service[name]);
+		});
+
+		return this;
+	}
```

Comment:
I'm a bit unsure about this. The README mentions that service references become static when registered in ifrau, but I'm not super sure where that happens. I worry that some of these arrow function conversions might have changed that behaviour, but tracking that down has proven to be a bit of a PITA because of the way the client/host communicate.

Reply:
`handler.apply(handler, ...)`

https://github.com/Brightspace/ifrau/pull/133/files/19e6673db89a7637585263902ff2fb92b3e70719#diff-e0791738063292273a16ec2305bb644dd086e4642bbac3124a7229c004d5f2bfR119

Reply:
Oh perfect, thanks! That makes more sense and makes me feel much less nervous about this refactor. :)

Reply:
Yeah, anything that's getting passed to `onRequest` would change behaviour when converted to an arrow function, if they were referencing `this`.

```
> const func = (function() { return () => this.x; }).call({ x: 1 });
> func.x = 2;
> const func2 = (function() { return function() { return this.x; }; }).call({ x: 3 });
> func2.x = 4;

> func()
1
> func.call(func)
1
> func2()
undefined
> func2.call(func2)
4
```